Docker FMU build fails with external c functions

### Description
When building an FMU using docker (w/ CMake or autoconf), external c functions cause the build to fail.  As @AnHeuermann observed in https://github.com/OpenModelica/OpenModelica/issues/9331#issuecomment-1447924577: 

- Suitability of multiarch/crossbuild: CMake version is too old.
- Adding external libraries locations to Docker container.
- Cross-Compilation for external libraries that don't provide the correct binary for OS inside Docker.

### Steps to reproduce
Create script `run.mos`:

```
loadString("
model externalCall
  Real x(start=1, fixed=true);
  Integer s = Modelica.Math.Random.Utilities.automaticGlobalSeed();
equation
  der(x) = -0.5*x;
end externalCall;"); getErrorString();
setCommandLineOptions("--fmuRuntimeDepends=none"); getErrorString();
setCommandLineOptions("--fmuCMakeBuild=true"); getErrorString();
buildModelFMU(externalCall,"2.0","cs","<default>",{"x86_64-linux-gnu docker run --pull=never multiarch/crossbuild"},true); getErrorString();
```
Running `omc run.mos` produces:

with cmake:
```
-- Linking ModelicaExternalC-NOTFOUND
-- Configuring incomplete, errors occurred!
```

without cmake (by commenting out the cmake related lines in run.mos):
```
/usr/bin/ld: cannot find -lModelicaExternalC
collect2: error: ld returned 1 exit status
```
### Expected behavior
FMU should build.

### Version and OS
 - latest OM / docker

### Additional context:
This question was originally posed here: https://github.com/OpenModelica/OpenModelica/issues/9331#issuecomment-1447344885, and @AnHeuermann asked to open a new issue.

--- a/OMCompiler/Compiler/Script/CevalScriptBackend.mo
+++ b/OMCompiler/Compiler/Script/CevalScriptBackend.mo
@@ -3735,7 +3735,6 @@ protected function configureFMU
   input String platform;
   input String fmutmp;
   input String logfile;
-  input list<String> externalLibLocations;
   input Boolean isWindows;
   input Boolean needs3rdPartyLibs;
 protected
@@ -3994,7 +3993,6 @@ protected
   String fmuTargetName;
   GlobalScript.SimulationOptions defaulSimOpt;
   SimCode.SimulationSettings simSettings;
-  SimCode.SimCode simCode;
   list<String> libs;
   Boolean isWindows;
   Boolean useCrossCompileCmake = false;
@@ -4118,7 +4116,7 @@ algorithm
     if useCrossCompileCmake then
       configureFMU_cmake(platform, fmutmp, filenameprefix, configureLogFile, libs, isWindows);
     else
-      configureFMU(platform, fmutmp, configureLogFile, libs, isWindows, needs3rdPartyLibs);
+      configureFMU(platform, fmutmp, configureLogFile, isWindows, needs3rdPartyLibs);
     end if;
     if Flags.getConfigEnum(Flags.FMI_FILTER) == Flags.FMI_BLACKBOX or Flags.getConfigEnum(Flags.FMI_FILTER) == Flags.FMI_PROTECTED then
       System.removeFile(configureLogFile);

--- a/OMCompiler/Compiler/Script/CevalScriptBackend.mo
+++ b/OMCompiler/Compiler/Script/CevalScriptBackend.mo
@@ -3539,6 +3539,7 @@ protected function configureFMU_cmake
   input String fmutmp;
   input String fmuTargetName;
   input String logfile;
+  input list<String> externalLibLocations;
   input Boolean isWindows;
 protected
   String fmuSourceDir;
@@ -3573,6 +3574,7 @@ algorithm
       Integer uid;
       String cidFile, volumeID, containerID, userID;
       String dockerLogFile;
+      list<String> locations, libraries;
     case {"dynamic"}
       algorithm
         if isWindows then
@@ -3600,6 +3602,10 @@ algorithm
 
         // Temp log file outside of Docker volume
         dockerLogFile := crossTriple + ".tmp.log";
+        // Remove old log file
+        if System.regularFileExists(dockerLogFile) then
+          System.removeFile(dockerLogFile);
+        end if;
 
         // Create a docker volume for the FMU since we can't forward volumes
         // to the docker run command depending on where the FMU was generated (inside another volume)
@@ -3623,6 +3629,17 @@ algorithm
         // Copy the FMI headers to the container
         cmd := "docker cp " + defaultFmiIncludeDirectoy + " " + containerID + ":/data/fmiInclude";
         runDockerCmd(cmd, dockerLogFile, cleanup=true, volumeID=volumeID, containerID=containerID);
+
+        // Copy the external library files to the container
+        (locations, libraries) := SimCodeUtil.getDirectoriesForDLLsFromLinkLibs(externalLibLocations);
+        print("libs:\n");
+        for loc in locations loop
+          print(loc + "\n");
+          if System.directoryExists(loc) then
+            cmd := "docker cp " + dquote + loc + dquote + " " + containerID + ":" + dquote + loc + dquote;
+            runDockerCmd(cmd, dockerLogFile, cleanup=true, volumeID=volumeID, containerID=containerID);
+          end if;
+        end for;
 
         // Build for target host
         userID := (if uid<>0 then "--user " + String(uid) else "");
@@ -3715,6 +3732,7 @@ protected function configureFMU
   input String platform;
   input String fmutmp;
   input String logfile;
+  input list<String> externalLibLocations;
   input Boolean isWindows;
   input Boolean needs3rdPartyLibs;
 protected
@@ -3973,6 +3991,7 @@ protected
   String fmuTargetName;
   GlobalScript.SimulationOptions defaulSimOpt;
   SimCode.SimulationSettings simSettings;
+  SimCode.SimCode simCode;
   list<String> libs;
   Boolean isWindows;
   Boolean useCrossCompileCmake = false;
@@ -4094,9 +4113,9 @@ algorithm
   for platform in platforms loop
     configureLogFile := System.realpath(fmutmp)+"/resources/"+System.stringReplace(listGet(Util.stringSplitAtChar(platform," "),1),"/","-")+".log";
     if useCrossCompileCmake then
-      configureFMU_cmake(platform, fmutmp, filenameprefix, configureLogFile, isWindows);
+      configureFMU_cmake(platform, fmutmp, filenameprefix, configureLogFile, libs, isWindows);
     else
-      configureFMU(platform, fmutmp, configureLogFile, isWindows, needs3rdPartyLibs);
+      configureFMU(platform, fmutmp, configureLogFile, libs, isWindows, needs3rdPartyLibs);
     end if;
     if Flags.getConfigEnum(Flags.FMI_FILTER) == Flags.FMI_BLACKBOX or Flags.getConfigEnum(Flags.FMI_FILTER) == Flags.FMI_PROTECTED then
       System.removeFile(configureLogFile);

--- a/OMCompiler/Compiler/Script/CevalScriptBackend.mo
+++ b/OMCompiler/Compiler/Script/CevalScriptBackend.mo
@@ -3632,11 +3632,13 @@ algorithm
 
         // Copy the external library files to the container
         (locations, libraries) := SimCodeUtil.getDirectoriesForDLLsFromLinkLibs(externalLibLocations);
-        print("libs:\n");
         for loc in locations loop
-          print(loc + "\n");
           if System.directoryExists(loc) then
-            cmd := "docker cp " + dquote + loc + dquote + " " + containerID + ":" + dquote + loc + dquote;
+            // Create path
+            cmd := "docker run --rm --hostname=" + containerID + " --volume=" + volumeID + ":/data busybox mkdir -p " + dquote + "/data" + loc + dquote;
+            runDockerCmd(cmd, dockerLogFile, cleanup=true, volumeID=volumeID, containerID=containerID);
+            // Copy files
+            cmd := "docker cp -a -L " + dquote + loc + dquote + " " + containerID + dquote + ":/data" + System.dirname(loc)  + dquote;
             runDockerCmd(cmd, dockerLogFile, cleanup=true, volumeID=volumeID, containerID=containerID);
           end if;
         end for;
@@ -3652,6 +3654,7 @@ algorithm
           fmiTarget := "";
         end if;
         cmakeCall := "cmake -DFMI_INTERFACE_HEADER_FILES_DIRECTORY=/fmu/fmiInclude " +
+                            "-DDOCKER_VOL_DIR=/fmu " +
                             fmiTarget +
                             CMAKE_BUILD_TYPE +
                             " ..";

--- a/OMCompiler/Compiler/SimCode/SimCodeUtil.mo
+++ b/OMCompiler/Compiler/SimCode/SimCodeUtil.mo
@@ -15662,14 +15662,14 @@ public function getCmakeLinkLibrariesCode
 protected
   list<String> locations;
   list<String> libraries;
-  function addQuotationMarks
+  function addDockerVol
     input String istring;
-    output String ostring = "\""+istring+"\"";
-  end addQuotationMarks;
+    output String ostring = "\"${DOCKER_VOL_DIR}"+istring+"\"";
+  end addDockerVol;
 algorithm
   (locations, libraries) := getDirectoriesForDLLsFromLinkLibs(libs);
   locations := listAppend({Settings.getInstallationDirectoryPath() + "/bin"}, locations);   // pthread located in OpenModelica/bin/ on Windows
-  locations := List.map(locations, addQuotationMarks);
+  locations := List.map(locations, addDockerVol);
   // Use target_link_directories when CMake 3.13 is available and skip the find_library part
   cmakecode := cmakecode + "set(EXTERNAL_LIBDIRECTORIES " + stringDelimitList(locations, "\n                            ") + ")\n";
   for lib in libraries loop

--- a/OMCompiler/Compiler/SimCode/SimCodeUtil.mo
+++ b/OMCompiler/Compiler/SimCode/SimCodeUtil.mo
@@ -15625,7 +15625,7 @@ algorithm
   end matchcontinue;
 end generateRunnerBatScript;
 
-protected function getDirectoriesForDLLsFromLinkLibs
+function getDirectoriesForDLLsFromLinkLibs
   " Parse the makefileParams.libs and extract the necessary directories
    to be added to the PATH for finding dependenncy libraries (DLLs on Windows).
    NOTE: this function expects the 'link command' as the second argument. This will

